### PR TITLE
[ADAM-1661] Add file storage benchmarks.

### LINKED ISSUE #1661: Add performance benchmarks similar to Samtools CRAM benchmarking page
Add performance benchmarks similar to http://www.htslib.org/benchmarks/CRAM.html on the same data.

Although I'm not sure these particular benchmarks actually do anything.  Can anyone translate?

"Conversion from BAM to is achieved via htslib test_view -b or -C. Decoding times are computed using test_view -B (benchmarking mode) with performs input, uncompression and decoding only. Encoding time includes decoding of the input BAM, so subtract the BAM decoding time to get encode-only, although the difference is only minor."

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1661. Is based on changes in #1771.

### CODE REVIEW

File: docs/source/35_benchmarks.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+In this benchmark, we have stored a copy of NA12878 aligned to the GRCh37
+reference genome using BWA. We store this genome in BAM, CRAM, and ADAM, using
+the default compression settings for each. BAM and CRAM files were generated
+using htslib. This read file was sequenced at approximately 60x coverage across
+the whole genome.
+
+![Storage cost of a 60x coverage WGS aligned dataset](source/img/bam.pdf)
+
+ADAM provides a 20% improvement in storage size over BAM, while CRAM achieves
+a 43% improvement in storage cost. While CRAM achieves a higher compression
```
Comment (heuermh) [inline docs/source/35_benchmarks.md]: `while CRAM ... While CRAM` reads a bit awkward

File: docs/source/35_benchmarks.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+Here, we benchmark both the GFF3 and BED formats. For GFF3, we use the ENSEMBL
+GRCh38 genome annotation file. For BED, we use genome-wide coverage counts
+generated from the NA12878 dataset used in the [aligned read
+benchmarks](#aligned-reads-storage).
+
+![Storage cost of genome annotations](source/img/gff.pdf)
+
+For the genome annotation file, ADAM provides a 20% improvement in storage size
+relative to the compressed GFF file.
```
Comment (heuermh) [inline docs/source/35_benchmarks.md]: GFF &rarr; GFF3

File: docs/source/img/source/file_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+# sizes in MB
+gff_sizes  = [  29.9,       36, 392.2]
+gff_labels = ['ADAM', 'GFF.GZ', 'GFF']
```
Comment (heuermh) [inline docs/source/img/source/file_benchmarks.py]: GFF &rarr; GFF3

File: docs/source/img/source/file_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    for (rect, size) in zip(rects, sizes):
+        height = rect.get_height()
+        ax.text(rect.get_x() + rect.get_width() / 2.0, 1.01 * height,
+                '%d %s' % (size, unit),
+                ha='center', va='bottom')
+    
+    fig.savefig(filename)
+
+plot(gff_sizes, gff_labels, 'gff.pdf', 'File sizes for ENSEMBL GRCh38 GTF', unit='MB')
```
Comment (heuermh) [inline docs/source/img/source/file_benchmarks.py]: GTF &rarr; GFF3

**Comment by fnothaft** [comment]
Jenkins, test this please.

**Comment by heuermh** [comment]
wot
```
VariantContextRDDSuite:
- union two variant context rdds together
- can write, then read in .vcf file
- can write as a single file via simple saveAsVcf method, then read in .vcf file
- can write as a single file via full saveAsVcf method, then read in .vcf file
- transform a vcf file with bad header
- read a vcf file with multi-allelic variants to split
- support VCFs with +Inf/-Inf float values *** FAILED ***
  java.lang.UnsupportedOperationException: empty collection
```

**Comment by fnothaft** [comment]
Jenkins, test this please.

**Comment by heuermh** [comment]
Thank you, @fnothaft

**Pull request merged**